prometheus-metrics-ghc/src/Prometheus/Metric/GHC.hs

```diff
@@ -125,13 +134,193 @@ ghcCollectors = [
+            max_live_bytes
+    , statsCollector
+            "ghc_max_large_objects_bytes"
+            "Maximum live data in large objects"
+            CounterType
+            max_large_objects_bytes
+    , statsCollector
+            "ghc_max_compact_bytes"
+            "Maximum live data in compact regions"
+            CounterType
+            max_compact_bytes
+    , statsCollector
+            "ghc_max_slop_bytes"
+            "Maximum slop"
+            CounterType
+            max_slop_bytes
+    , statsCollector
+            "ghc_max_mem_in_use_bytes"
+            "Maximum memory in use by the RTS"
+            CounterType
+            max_mem_in_use_bytes
+    , statsCollector
+            "ghc_cumulative_live_bytes_total"
+            "Sum of live bytes across all major GCs. Divided by major_gcs gives the average live data over the lifetime of the program."
+            CounterType
+            cumulative_live_bytes
+    , statsCollector
+            "ghc_copied_bytes_total"
+            "Sum of copied_bytes across all GCs"
+            CounterType
+            copied_bytes
+    , statsCollector
+            "ghc_par_copied_bytes_total"
+            "Sum of copied_bytes across all parallel GCs"
+            CounterType
+            par_copied_bytes
+    , statsCollector
+            "ghc_cumulative_par_max_copied_bytes_total"
+            "Sum of par_max_copied_bytes across all parallel GCs"
+            CounterType
+            cumulative_par_max_copied_bytes
+    , statsCollector
+            "ghc_mutator_cpu_seconds_total"
+            "Total CPU time used by the mutator"
+            CounterType
+            (rtsTimeToSeconds . mutator_cpu_ns)
+    , statsCollector
+            "ghc_mutator_elapsed_seconds_total"
+            "Total elapsed time used by the mutator"
+            CounterType
+            (rtsTimeToSeconds . mutator_elapsed_ns)
+    , statsCollector
+            "ghc_gc_cpu_seconds_total"
+            "Total CPU time used by the GC"
+            CounterType
+            (rtsTimeToSeconds . gc_cpu_ns)
+    , statsCollector
+            "ghc_gc_elapsed_seconds_total"
+            "Total elapsed time used by the GC"
+            CounterType
+            (rtsTimeToSeconds . gc_elapsed_ns)
+    , statsCollector
+            "ghc_cpu_seconds_total"
+            "Total CPU time (at the previous GC)"
+            CounterType
+            (rtsTimeToSeconds . cpu_ns)
+    , statsCollector
+            "ghc_elapsed_seconds_total"
+            "Total elapsed time (at the previous GC)"
+            CounterType
+            (rtsTimeToSeconds . elapsed_ns)
+
+    , statsCollector
+            "ghc_gcdetails_gen"
+            "The generation number of this GC"
+            HistogramType -- TODO: is this correct?
+                          -- Gauge makes little sense here.
+                          -- With Histogram we'll be able to see which
+                          -- generations are collected most often.
```

Comment:
Histogram would be useful if we wanted to know the median generation or something like that. I don't know enough about GHC's GC to say what makes sense.

e.g. perhaps it would be useful to have a counter with a label per generation. Perhaps other of these metrics would benefit from having the generation as a label?

For the histogram to work well, you'd want to be careful about bucket selection.

Reply:
> perhaps it would be useful to have a counter with a label per generation

That sounds like a great idea. But that would be a bigger change. Would it be OK if I just made this a `Gauge` for now?

Reply:
Looking over this again, I think the right thing to do is for this value to be a label on all of these other metrics. I’m not 100% certain of this, however.

Making this a gauge won’t hurt, and doesn’t set us up for difficult compatibility problems later, so it’s fine by me.